src/edu/washington/escience/myria/operator/SampledDbInsertTemp.java

```diff
@@ -0,0 +1,221 @@
+      ImmutableList.of(Type.INT_TYPE, Type.INT_TYPE, Type.INT_TYPE),
+      ImmutableList.of("WorkerID", "PartitionSize", "StreamSize"));
+
+  /**
+   * @param child
+   *          the source of tuples to be inserted.
+   * @param streamSampleSize
+   *          number of tuples to store from the stream
+   * @param sampleRelationKey
+   *          the key of the table that the tuples should be inserted into.
+   * @param countRelationKey
+   *          the key of the table that the tuple counts will be inserted into.
+   * @param connectionInfo
+   *          the parameters of the database connection.
+   */
+  public SampledDbInsertTemp(final Operator child, final int streamSampleSize,
+      final RelationKey sampleRelationKey, final RelationKey countRelationKey,
+      final ConnectionInfo connectionInfo) {
+    super(child);
+    // Sampling setup.
+    Preconditions.checkArgument(streamSampleSize >= 0L,
+        "sampleSize must be non-negative");
+    this.streamSampleSize = streamSampleSize;
+    doneSamplingFromChild = false;
+
+    // Relation setup.
+    Objects.requireNonNull(sampleRelationKey, "sampleRelationKey");
+    this.sampleRelationKey = sampleRelationKey;
+    Objects.requireNonNull(countRelationKey, "countRelationKey");
+    this.countRelationKey = countRelationKey;
+    this.connectionInfo = connectionInfo;
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws DbException {
+    if (!doneSamplingFromChild) {
+      fillReservoir();
+      batches = reservoir.getAll();
+      // Insert sampled tuples into sampleRelationKey
+      while (batchNum < batches.size()) {
+        TupleBatch batch = batches.get(batchNum);
+        accessMethod.tupleBatchInsert(sampleRelationKey, batch);
+        batchNum++;
+      }
+
+      // Write (WorkerID, PartitionSize, StreamSize) to countRelationKey
+      IntColumnBuilder wIdCol = new IntColumnBuilder();
+      IntColumnBuilder tupCountCol = new IntColumnBuilder();
+      IntColumnBuilder streamSizeCol = new IntColumnBuilder();
+      wIdCol.appendInt(getNodeID());
+      tupCountCol.appendInt(tupleCount);
+      streamSizeCol.appendInt(reservoir.numTuples());
+      ImmutableList.Builder<Column<?>> columns = ImmutableList.builder();
+      columns.add(wIdCol.build(), tupCountCol.build(), streamSizeCol.build());
+      TupleBatch tb = new TupleBatch(COUNT_SCHEMA, columns.build());
+      accessMethod.tupleBatchInsert(countRelationKey, tb);
+    }
+    return null;
+  }
+
+  /**
+   * Fills reservoir with child tuples.
+   *
+   * @throws DbException
+   *           if TupleBatch fails to get nextReady
+   */
+  private void fillReservoir() throws DbException {
+    Random rand = new Random();
+    for (TupleBatch tb = getChild().nextReady(); tb != null; tb = getChild()
+        .nextReady()) {
+      final List<? extends Column<?>> columns = tb.getDataColumns();
+      for (int i = 0; i < tb.numTuples(); i++) {
+        if (reservoir.numTuples() < streamSampleSize) {
```

Comment:
Reservoir sampling is much better-known than Floyd's sampling algorithm, but it might still be helpful to add a reference (Wikipedia article is fine), so the reader knows exactly which algorithm is being used.